Signature: Move HTTP request signing to a filter

obenland opened the pull request:
Follow up to #1986 and #1991.

Moves double-knock into a filter that's always set.
With `Signature` getting an init callback, we can now also move the signing logic into a hook callback.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Moved HTTP request signing to a filter in Signature::init and removed direct calls from Http class. 
* Updated Signature class to add filters for signing and double knock logic. 
* Adjusted tests to use filter-based approach for signature verification.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [x] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Moved HTTP request signing to a filter instead of calling it directly.

</details>

obenland committed:
Signature: Move HTTP request signing to a filter

Moved HTTP request signing to a filter in Signature::init and removed direct calls from Http class. Updated Signature class to add filters for signing and double knock logic. Adjusted tests to use filter-based approach for signature verification.

matticbot committed:
Add changelog

pfefferle commented:
You remove the `method` from the `$args` but it seems that `Http_Signature_Draft` is still using it in line 49.

I wonder why the tests are green 🤔

pfefferle commented:
Ah, the test is still adding the `method` in the `$args` array!

obenland commented:
Yes, that would be expected. Tests create signatures manually to sign "incoming" requests and they need the method provided.

When signing outgoing requests, `http_request_args` has a method set in those request args, [`GET` by default](https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/class-wp-http.php#L171), and any specific method based on the type of request is sent ([post](https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/class-wp-http.php#L636), [head](https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/class-wp-http.php#L672) ).

obenland committed:
Merge branch 'trunk' into update/signature-callbacks

pfefferle commented:
nice job!

pfefferle committed:
Merge branch 'trunk' into update/signature-callbacks

pfefferle committed:
Merge branch 'trunk' into update/signature-callbacks

obenland committed:
Refactor signature tests to use http_request_args filter

Updated signature-related tests to apply the 'http_request_args' filter directly instead of invoking the sign() method. Also added setup and cleanup for the 'activitypub_rfc9421_signature' option in the RFC9421 test to ensure proper test isolation.

obenland committed:
Merge branch 'trunk' into update/signature-callbacks

obenland committed:
Merge branch 'trunk' into update/signature-callbacks
